taichi/program/sparse_solver.cpp

```diff
@@ -32,17 +32,37 @@ bool EigenSparseSolver<EigenSolver>::info() {
   return solver_.info() == Eigen::Success;
 }
 
-std::unique_ptr<SparseSolver> get_sparse_solver(
-    const std::string &solver_type) {
-  if (solver_type == "LU") {
+#include <map>
+using key_type = std::pair<std::string, std::string>;
+using func_type = std::unique_ptr<SparseSolver> (*)();
+
+#define MAKE_SOLVER(type, order)                                              \
+  {                                                                           \
+    {#type, #order}, []() -> std::unique_ptr<SparseSolver> {                  \
+      using T =                                                               \
+          Eigen::Simplicial##type<Eigen::SparseMatrix<float32>, Eigen::Lower, \
+                                  Eigen::order##Ordering<int>>;               \
+      return std::make_unique<EigenSparseSolver<T>>();                        \
+    }                                                                         \
+  }
+
+static std::map<key_type, func_type> solver_factory = {
```

Comment:
nit: move this into `get_sparse_solver` and make it const? use unordered_map

Reply:
OK, I could move `solver_factory` into  `get_sparse_solver`.

As far as I know, `unordered_map` couldn't hash `std::pair`.

Reply:
Yeah, you can provide a customized hasher object for that specific type of pair. See the third template type of https://en.cppreference.com/w/cpp/container/unordered_map :-)